### PR TITLE
Watchdog timer change v1

### PR DESCRIPTION
**Pull request by kmk3210**
This patch is about supporting api which changes "WATCHDOG_USEC" value during service running.
With this api, running service can change its own watchdog timeout value.

Detail discussions can be referenced from this link : https://lists.freedesktop.org/archives/systemd-devel/2016-May/036552.html

### CODE REVIEW

File: src/core/service.c [hunk 1]
```diff
@@ -202,14 +202,24 @@ static void service_stop_watchdog(Service *s) {
...
+        if (s->watchdog_usec == 0 || s->watchdog_usec == USEC_INFINITY)
                 return;
 
+        if (s->watchdog_override_usec == USEC_INFINITY)
+                 return;
+
+        if (s->watchdog_override_usec != 0)
+            watchdog_usec = s->watchdog_override_usec;
+        else
+            watchdog_usec = s->watchdog_usec;
```
Comment (poettering) [inline src/core/service.c]: Indentation is borked. Please see CODING_STYLE: 8ch indentation please!

Also, could you please create a new function for this: service_get_watchdog_usec() which returns the current watchdog interval, either taking the override value or the original value?

File: src/core/service.c [hunk 2]
```diff
@@ -246,6 +256,20 @@ static void service_reset_watchdog(Service *s) {
         service_start_watchdog(s);
 }
 
+static void service_reset_watchdog_timeout(Service *s, usec_t watchdog_override_usec) {
+        assert(s);
+
+        if (watchdog_override_usec == 0)
+                return ;
```
Comment (poettering) [inline src/core/service.c]: I am pretty sure there should be a way to fully disable the watchdog, but this makes that impossible no?

File: src/core/service.c [hunk 3]
```diff
@@ -246,6 +256,20 @@ static void service_reset_watchdog(Service *s) {
...
+
+        s->watchdog_override_usec = watchdog_override_usec;
+        service_reset_watchdog(s);
+
+        log_unit_debug(UNIT(s), "watchdog timeout :  "USEC_FMT, s->watchdog_usec);
+        log_unit_debug(UNIT(s), "override watchdog timeout :  "USEC_FMT, s->watchdog_override_usec);
```
Comment (poettering) [inline src/core/service.c]: weird spaces in the format string...

File: src/core/service.c [hunk 4]
```diff
@@ -246,6 +256,20 @@ static void service_reset_watchdog(Service *s) {
...
+}
+
+
 static void service_fd_store_unlink(ServiceFDStore *fs) {
```
Comment (poettering) [inline src/core/service.c:249]: spurious newline

File: src/core/service.c [hunk 5]
```diff
@@ -2134,6 +2160,9 @@ static int service_serialize(Unit *u, FILE *f, FDSet *fds) {
 
         unit_serialize_item(u, f, "forbid-restart", yes_no(s->forbid_restart));
 
+        if (s->watchdog_override_usec)
+               unit_serialize_item_format(u, f, "watchdog_override_usec", USEC_FMT, s->watchdog_override_usec);
```
Comment (poettering) [inline src/core/service.c]: all other serialized props use "-" instead of "_" as word separator, please do the same.

File: src/core/service.c [hunk 6]
```diff
@@ -2134,6 +2160,9 @@ static int service_serialize(Unit *u, FILE *f, FDSet *fds) {
 
         unit_serialize_item(u, f, "forbid-restart", yes_no(s->forbid_restart));
 
+        if (s->watchdog_override_usec)
```
Comment (poettering) [inline src/core/service.c]: nitpick: only bools and ptrs we check with the degrade-to-bool feature of C, for everything numeric please use numeric checks. Or in this case: please write this as `if (s->watchdog_override_usec != 0)`

File: src/core/service.c [hunk 7]
```diff
@@ -3047,6 +3090,15 @@ static void service_notify_message(Unit *u, pid_t pid, char **tags, FDSet *fds)
                 service_add_fd_store_set(s, fds, name);
         }
 
+        e = strv_find_startswith(tags, "WDOGSEC=");
```
Comment (poettering) [inline src/core/service.c]: please just call this WATCHDOG_USEC= the same as the env var. There's no need to abbreviate this, and it certainly doesn't help readability.

**Comment by poettering** [comment]
So, I am not sure about the WatchdogMaxSec= thing. I doubt it brings any benefits and suffers by a major design problem. The design problem is that there's no useful way for services that set watchdog seconds above the limit to know that what they just set didn't work. And if they don't know that they are likely to be killed pretty soon by the watchdog handler.

I am also not sure what security benefit WatchdogMaxSec= should provide. I mean, if a service really wants to escape watchdog supervision it may trivially fork off a thread, which keeps pinging PID 1... 

hence I think the second patch really should just go, it's not necessary, it's conceptually flawed.

**Comment by poettering** [comment]
The first patch looks conceptually OK, I added some comments though.

File: src/core/service.c (hunk 2 above)
Comment (kmk3210) [inline src/core/service.c]: I thought that changed watchdog timeout for 0 sec is meaningless,
but now I think I need to change it.
I will add new variable watchdog_override_enable, so watchdog_override_usec does not need having range.

File: src/core/service.c (hunk 1 above)
Comment (kmk3210) [inline src/core/service.c]: Just for internal function? I'm gonna make it.!

**Pull request closed without merging**